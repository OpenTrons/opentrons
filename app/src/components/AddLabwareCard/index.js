// @flow
import * as React from 'react'
import { Card } from '@opentrons/components'

// TODO(mc, 2019-10-17): i18n
const ADD_LABWARE_CARD_TITLE = 'Labware Management'

function AddLabwareCard() {
  return <Card title={ADD_LABWARE_CARD_TITLE}></Card>
}

export default AddLabwareCard
