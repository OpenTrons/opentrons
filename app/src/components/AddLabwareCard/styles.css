@import '@opentrons/components';

.labeled_value {
  margin-top: 0.25rem;
  margin-bottom: 0;
}

.code {
  @apply --font-code-dark;
}

.details_list {
  list-style: none;
  margin-bottom: 1rem;
}

.list_item {
  display: flex;
  align-items: baseline;
  margin-top: 0.125rem;
}

.list_item_title {
  font-weight: var(--fw-semibold);
  margin-right: 0.25rem;
  min-width: 5rem;
}

.button {
  width: auto;
  padding-left: 2rem;
  padding-right: 2rem;
}

.error {
  font-weight: var(--fw-semibold);
  color: var(--c-error-dark);
}
