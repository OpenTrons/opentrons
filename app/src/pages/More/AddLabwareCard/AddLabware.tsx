import * as React from 'react'

import { LabeledButton } from '@opentrons/components'

// TODO(mc, 2019-11-18): i18n
const ADD_NEW_LABWARE_DEFINITIONS = 'Add New Labware Definitions'
const ADD_LABWARE = 'Add Labware'
const ADD_LABWARE_DESCRIPTION =
  'Add labware definitions to your Custom Labware Definitions folder for use on any OT-2 robot.'
const CUSTOM_LABWARE_CREATOR = 'Custom Labware Creator'
const USE_THE = 'Use the'
const TO_GENERATE_NEW_LABWARE_DEFINITIONS =
  'to generate new labware definitions'

const LABWARE_CREATOR_HREF = 'https://labware.opentrons.com/create'

export const ADD_LABWARE_NAME = 'add-labware'

export interface AddLabwareProps {
  onAddLabware: () => unknown
}

export function AddLabware(props: AddLabwareProps): JSX.Element {
  const { onAddLabware } = props

  return (
    <LabeledButton
      label={ADD_NEW_LABWARE_DEFINITIONS}
      buttonProps={{
        name: ADD_LABWARE_NAME,
        children: ADD_LABWARE,
        onClick: onAddLabware,
      }}
    >
      <p>{ADD_LABWARE_DESCRIPTION}</p>
      <p>
        {USE_THE}{' '}
        <a
          href={LABWARE_CREATOR_HREF}
          target="_blank"
          rel="noopener noreferrer"
        >
          {CUSTOM_LABWARE_CREATOR}
        </a>{' '}
        {TO_GENERATE_NEW_LABWARE_DEFINITIONS}
      </p>
    </LabeledButton>
  )
}
