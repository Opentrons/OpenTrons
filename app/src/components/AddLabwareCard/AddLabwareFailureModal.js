// @flow
import * as React from 'react'

import { AlertModal } from '@opentrons/components'
import { Portal } from '../portal'
import styles from './styles.css'

import {
  INVALID_LABWARE_FILE,
  DUPLICATE_LABWARE_FILE,
  OPENTRONS_LABWARE_FILE,
} from '../../redux/custom-labware'

import type {
  FailedLabwareFile,
  DuplicateLabwareFile,
} from '../../redux/custom-labware/types'

// TODO(mc, 2019-11-20): i18n
// buttons
const CANCEL = 'cancel'
const OVERWRITE_LABWARE = 'overwrite labware'

// headings
const UNABLE_TO_ADD_LABWARE = 'Unable to add labware'
const INVALID_LABWARE_DEFINITION = 'Invalid labware definition'
const CONFLICT_WITH_OPENTRONS_LABWARE = 'Conflict with Opentrons labware'
const OVERWRITE_DUPLICATE_LABWARE = 'Overwrite duplicate labware?'

// copy
const UNABLE_TO_COPY_LABWARE_FILE_TO = 'Unable to copy labware file to'
const PLEASE_CHECK_PERMISSIONS =
  'Please check that you are able to open this file and add files to your labware source folder, or contact Opentrons Support for help.'
const THE_FILE = 'The file'
const IS_NOT_A_VALID_LABWARE_DEFINITION =
  'is not a valid Opentrons labware definition. Please check that you selected to correct file.'
const THIS_LABWARE_DEFINITION_CONFLICTS =
  'This labware definition conflicts with an Opentrons standard definition. If you are trying to create a new labware based on an Opentrons definition, please contact support.'
const A_LABWARE_DEFINITION_ALREADY_EXISTS =
  'A labware definition with the same API name, version, and namespace already exists in your labware source folder. Would you like to overwrite it?'
const CLICKING_OVERWRITE_LABWARE_WILL_DELETE_FILES =
  'Clicking "Overwrite Labware" will delete any existing files that conflict with the new file'

const DISPLAY_NAME = 'Name: '
const API_NAME = 'API name: '
const FILENAME = 'File name: '

export type AddLabwareFailureModalProps = {|
  file: FailedLabwareFile | null,
  errorMessage: string | null,
  directory: string,
  onCancel: () => mixed,
  onOverwrite: (file: DuplicateLabwareFile) => mixed,
|}

const renderFilename = file => (
  <span className={styles.code}>{file.filename}</span>
)

const renderDetails = file => (
  <>
    <ul className={styles.details_list}>
      <li className={styles.list_item}>
        <span className={styles.list_item_title}>{DISPLAY_NAME}</span>
        <span>{file.definition.metadata.displayName}</span>
      </li>
      <li className={styles.list_item}>
        <span className={styles.list_item_title}>{API_NAME}</span>
        <span className={styles.code}>
          {file.definition.parameters.loadName}
        </span>
      </li>
      <li className={styles.list_item}>
        <span className={styles.list_item_title}>{FILENAME}</span>
        <span>{renderFilename(file)}</span>
      </li>
    </ul>
  </>
)

export function AddLabwareFailureModalTemplate(
  props: AddLabwareFailureModalProps
): React.Node {
  const { file, errorMessage, directory, onCancel, onOverwrite } = props
  let buttons = [
    { onClick: onCancel, children: CANCEL, className: styles.button },
  ]
  let heading = null
  let children = null

  if (!file || errorMessage) {
    heading = UNABLE_TO_ADD_LABWARE
    children = (
      <>
        <p>
          {UNABLE_TO_COPY_LABWARE_FILE_TO}{' '}
          <span className={styles.code}>{directory}</span>
          {'. '}
          {PLEASE_CHECK_PERMISSIONS}
        </p>
        {errorMessage && <p className={styles.error}>{errorMessage}</p>}
      </>
    )
  } else if (file.type === INVALID_LABWARE_FILE) {
    heading = INVALID_LABWARE_DEFINITION
    children = (
      <>
        <p>
          {THE_FILE} {renderFilename(file)} {IS_NOT_A_VALID_LABWARE_DEFINITION}
        </p>
        {file.errors &&
          file.errors.map(({ dataPath, message }) => (
            <p>
              {dataPath} {message}
            </p>
          ))}
      </>
    )
  } else if (file.type === OPENTRONS_LABWARE_FILE) {
    heading = CONFLICT_WITH_OPENTRONS_LABWARE
    children = (
      <>
        <p>{THIS_LABWARE_DEFINITION_CONFLICTS}</p>
        {renderDetails(file)}
      </>
    )
  } else if (file.type === DUPLICATE_LABWARE_FILE) {
    buttons = [
      ...buttons,
      {
        className: styles.button,
        children: OVERWRITE_LABWARE,
        onClick: () => onOverwrite(file),
      },
    ]
    heading = OVERWRITE_DUPLICATE_LABWARE
    children = (
      <>
        <p>{A_LABWARE_DEFINITION_ALREADY_EXISTS}</p>
        {renderDetails(file)}
        <p>{CLICKING_OVERWRITE_LABWARE_WILL_DELETE_FILES}</p>
      </>
    )
  }

  return (
    <AlertModal alertOverlay heading={heading} buttons={buttons}>
      {children}
    </AlertModal>
  )
}

export function AddLabwareFailureModal(
  props: AddLabwareFailureModalProps
): React.Node {
  return (
    <Portal>
      <AddLabwareFailureModalTemplate {...props} />
    </Portal>
  )
}
