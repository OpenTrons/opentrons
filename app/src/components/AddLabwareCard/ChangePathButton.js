// @flow

import * as React from 'react'
import { OutlineButton } from '@opentrons/components'

import styles from './styles.css'

// TODO(mc, 2019-11-18): i18n
const CHANGE_SOURCE = 'Change Source'

export type ChangePathButtonProps = {|
  onChangePath: () => mixed,
|}

export function ChangePathButton(props: ChangePathButtonProps) {
  return (
    <OutlineButton
      className={styles.change_path_button}
      onClick={props.onChangePath}
    >
      {CHANGE_SOURCE}
    </OutlineButton>
  )
}
