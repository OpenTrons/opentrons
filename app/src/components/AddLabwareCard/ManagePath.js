// @flow

import * as React from 'react'

import { LabeledButton, InputField } from '@opentrons/components'
import styles from './styles.css'

// TODO(mc, 2019-11-18): i18n
const CUSTOM_LABWARE_DEFINITIONS_FOLDER = 'Custom Labware Definitions Folder'
const CHANGE_SOURCE = 'Change Source'

export type ManagePathProps = {|
  path: string,
  onChangePath: () => mixed,
|}

export function ManagePath(props: ManagePathProps) {
  const { path, onChangePath } = props
  const handleFocus = (e: SyntheticFocusEvent<HTMLInputElement>) => {
    e.currentTarget.select()
  }

  return (
    <LabeledButton
      label={CUSTOM_LABWARE_DEFINITIONS_FOLDER}
      buttonProps={{
        onClick: onChangePath,
        children: CHANGE_SOURCE,
      }}
    >
      <InputField
        readOnly
        value={path}
        type="text"
        onFocus={handleFocus}
        className={styles.labeled_value}
      />
    </LabeledButton>
  )
}
