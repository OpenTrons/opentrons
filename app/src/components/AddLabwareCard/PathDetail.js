// @flow

import * as React from 'react'

import { InputField } from '@opentrons/components'
import styles from './styles.css'

// TODO(mc, 2019-11-18): i18n
const CUSTOM_LABWARE_DEFINITIONS_FOLDER = 'Custom Labware Definitions Folder'

export type PathDetailProps = {|
  path: string,
|}

export function PathDetail(props: PathDetailProps) {
  return (
    <div className={styles.path_detail}>
      <h4 className={styles.path_detail_title}>
        {CUSTOM_LABWARE_DEFINITIONS_FOLDER}
      </h4>
      <InputField
        readOnly
        value={props.path}
        type="text"
        onFocus={(e: SyntheticFocusEvent<HTMLInputElement>) => {
          e.currentTarget.select()
        }}
      />
    </div>
  )
}
