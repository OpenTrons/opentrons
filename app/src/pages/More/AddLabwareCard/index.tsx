import * as React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Card } from '@opentrons/components'

import {
  getCustomLabwareDirectory,
  openCustomLabwareDirectory,
  changeCustomLabwareDirectory,
  resetCustomLabwareDirectory,
  addCustomLabware,
  clearAddCustomLabwareFailure,
  getAddLabwareFailure,
} from '../../../redux/custom-labware'

import { CardCopy } from '../../../atoms/layout'
import { ManagePath } from './ManagePath'
import { AddLabware } from './AddLabware'
import { AddLabwareFailureModal } from './AddLabwareFailureModal'

import type { Dispatch } from '../../../redux/types'

// TODO(mc, 2019-10-17): i18n
const LABWARE_MANAGEMENT = 'Labware Management'
const MANAGE_CUSTOM_LABWARE_DEFINITIONS =
  'Manage custom labware definitions for use in your Python Protocol API Version 2 protocols.'

export function AddLabwareCard(): JSX.Element {
  const dispatch = useDispatch<Dispatch>()
  const labwarePath = useSelector(getCustomLabwareDirectory)
  const addFailure = useSelector(getAddLabwareFailure)
  const handleOpenPath = (): unknown => dispatch(openCustomLabwareDirectory())
  const handleResetPath = (): unknown => dispatch(resetCustomLabwareDirectory())
  const handleChangePath = (): unknown =>
    dispatch(changeCustomLabwareDirectory())
  const handleAddLabware = (): unknown => dispatch(addCustomLabware())
  const showAddFailure = addFailure.file || addFailure.errorMessage !== null

  return (
    <Card title={LABWARE_MANAGEMENT}>
      <CardCopy>{MANAGE_CUSTOM_LABWARE_DEFINITIONS}</CardCopy>
      <ManagePath
        path={labwarePath}
        onOpenPath={handleOpenPath}
        onResetPath={handleResetPath}
        onChangePath={handleChangePath}
      />
      <AddLabware onAddLabware={handleAddLabware} />
      {showAddFailure && (
        <AddLabwareFailureModal
          {...addFailure}
          directory={labwarePath}
          onCancel={() => dispatch(clearAddCustomLabwareFailure())}
          onOverwrite={file => dispatch(addCustomLabware(file))}
        />
      )}
    </Card>
  )
}
