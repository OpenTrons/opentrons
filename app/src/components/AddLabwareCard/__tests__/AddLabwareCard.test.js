// @flow
import * as React from 'react'
import { shallow } from 'enzyme'

import AddLabwareCard from '..'

describe('AddLabwareCard', () => {
  test('component renders', () => {
    const tree = shallow(<AddLabwareCard />)

    expect(tree).toMatchSnapshot()
  })
})
