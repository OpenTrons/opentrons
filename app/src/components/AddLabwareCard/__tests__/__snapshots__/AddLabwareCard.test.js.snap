// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLabwareCard component renders 1`] = `
<Card
  title="Labware Management"
/>
`;
